### PR TITLE
[[ Enhancement ]]: FilePicker - added  onFileLoaded and onFileDeselected events

### LINKED ISSUE #3672: [ FilePicker ]: 'On file loaded' and 'On file deselected' events are needed
### Is your proposal related to a problem?

1. When the user selects a file, the On file selected event is fired, however, this event is fired seconds before the selected file is loaded and made available on the application. 
2. There is no event fired when a file is deselected(deleted) from the file picker. 

### Describe the solution you'd like

1. So we need an event that is fired after the file is loaded and made available to be accessed.
2. We need to fire onFileDeselected event when deleting a file

### PR DESCRIPTION
**Pull request by manishkushare**
Resolve #3672

### CODE REVIEW

**Comment by arpitnath** [comment]
@manishkushare 
When we attach an action to the `On File Loaded` event, the action is not triggered. 
https://www.loom.com/share/e22ec42bb5dc4b27a87d9c20ffc1f28d

File: frontend/src/Editor/Components/FilePicker.jsx [hunk 1]
```diff
@@ -241,18 +241,20 @@ export const FilePicker = ({
...
+          return new Promise(function (resolve, reject) {
+            setTimeout(() => {
+              setShowSelectedFiles(true);
+              setAccepted(false);
+              onComponentOptionChanged(component, 'isParsing', false);
+              resolve();
+            }, 600);
+          });
+        })
+        .then(() => onEvent('onFileLoaded ', { component }));
```
Comment (arpitnath) [inline frontend/src/Editor/Components/FilePicker.jsx:257]: There is an extra white space at the end of the **eventName**, 
`frontend/src/_helpers/appUtils.js :: 488-518` return `false` and the action is not been executed.

**Review by arpitnath — CHANGES_REQUESTED** [review]

**Review by arpitnath — APPROVED** [review]
LGTM!

**Comment by Nandinisaha13** [comment]
<html>
<body>
<google-sheets-html-origin>Test Case

Test Case | Steps | Result
-- | -- | --
Should be able to show alert(info) while file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select alert type as info. | Alert type info is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | Alert(ifo) is displayed.
Should be able to show alert(sucess) while file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select alert type as success. | Alert type success is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | Alert(success) is displayed.
Should be able to show alert(warning) while file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select alert type as warning. | Alert type warning is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | Alert(warning) is displayed.
Should be able to show alert(danger) while file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select alert type as danger. | Alert type danger is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | Alert(danger) is displayed.
Should be able to logout when file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select logout as action. | Logout is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | User is logged out.
Should be able to run query when file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Create a query. | Query is created.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select run query as action & select the above query. | Run query is selected & the above query is selected in query field.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | The query is successfully executed.
Should be able to open webpage when file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select open webpage in action & enter the url. | Open webpage is selected & url is entered.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | The webpage is opened.
Should be able to open another application when file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select go to app in action & select app name. | Go to app is selected & app name is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | Pop up is opened for confirmation for triggering the application.
  | Click open. | Application is displayed.
Should be able to open modal when file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Drag & drop modal. | Modal is displayed.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select open modal & select modal name. | Open modal is selected & modal name is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | Modal is opened.
Should be able to close modal when file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Drag & drop modal. | Modal is displayed.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select open modal & select modal name. | Open modal is selected & modal name is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | Modal is opened.
  | Drag & drop fliepicker inside the modal. | Filepicker is displayed inside the modal.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select close modal & select modal name. | close modal is selected & modal name is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | Modal is closed.
Should be able to copy to clipboard when file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select copy to clipboard in action & enter text. | Copy to clipboard is selected & text is entered.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | Text is copied to clipboard.
Should be able to set local storage when file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Drag & drop text input & text widget. | Text & text input is displayed.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select set local storage in action & enter key & value. | Set local storage is selected & key & value is entered.
  | Enter {{localStorage.getItem('name')}}. | Entered.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | Text widget shows the text entered in text input.
  | Click on refresh. | The local storage is intact,
Should be able to generate file when file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select generate a file in action & enter file name & select type as csv & enter data. | Generate a file is selected & enter file name & csv is elected as type & data is entered,
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is generated as csv type.
Should be able to set table page when file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Drag & Drop table widget. | Table widget is displayed.
  | Enter more details in the table. | Details are entered.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select set table page & select table name & enter page number. | Set table page is selected & page number is entered.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | Table page is set to the page number selected.
Should be able to set variable when file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select set variable & enter name. | Set variable is selected in action & name is entered.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | Variable is set to the name.
  | Click on inspector. | Variable is displayed.
  | Click on variable. | The name is displayed.
Should be able to unset variable when file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select unset variable. | Unset variable is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | Variable is unset.
  | Click on inspector. | Variable with 0 items is displayed.
Should be able to control component when file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select control component & select compnent name & action. | Control component is selected & component name & action is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | Component is controlled as per action.
Should be able to show alert(info) while file is deselected. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select alert type as info. | Alert type info is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is selected. | File is selected on file picker.
  | Delete the file. | Alert(info) is displayed.
Should be able to show alert(success) while file is deleselected. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select alert type as success. | Alert type success is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is selected on file picker.
  | Delete the file. | Alert(success) is displayed.
Should be able to show alert(warning) while file is loaded. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select alert type as warning. | Alert type warning is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is selected on file picker.
  | Delete the file. | Alert(warning) is displayed.
Should be able to show alert(danger) while file is deselected. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select alert type as danger. | Alert type danger is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is selected on file picker.
  | Delete the file. | Alert(danger) is displayed.
Should be able to logout when file is deselected. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select logout as action. | Logout is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is selected on file picker.
  | Delete the file. | User is logged out.
Should be able to run query when file is deselected. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Create a query. | Query is created.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select run query as action & select the above query. | Run query is selected & the above query is selected in query field.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is selected on file picker.
  | Delete the file. | The query is successfully executed.
Should be able to open webpage when file is deslected. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select open webpage in action & enter the url. | Open webpage is selected & url is entered.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is selected on file picker.
  | Delete the file. | The webpage is opened.
Should be able to open another application when file is deselected. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select go to app in action & select app name. | Go to app is selected & app name is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is selected on file picker.
  | Delete the file. | Pop up is opened for confirmation for triggering the application.
  | Click open. | Application is displayed.
Should be able to open modal when file is deselected. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Drag & drop modal. | Modal is displayed.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select open modal & select modal name. | Open modal is selected & modal name is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is selected on file picker.
  | Delete the file. | Modal is opened.
Should be able to close modal when file is deselected. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Drag & drop modal. | Modal is displayed.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select open modal & select modal name. | Open modal is selected & modal name is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | Modal is opened.
  | Drag & drop fliepicker inside the modal. | Filepicker is displayed inside the modal.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select close modal & select modal name. | close modal is selected & modal name is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is selected on file picker.
  | Delete the file. | Modal is closed.
Should be able to copy to clipboard when file is deselected. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select copy to clipboard in action & enter text. | Copy to clipboard is selected & text is entered.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is selected on file picker.
  | Delete the file. | Text is copied to clipboard.
Should be able to set local storage when file is deselected. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Drag & drop text input & text widget. | Text & text input is displayed.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select set local storage in action & enter key & value. | Set local storage is selected & key & value is entered.
  | Enter {{localStorage.getItem('name')}}. | Entered.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is selected on file picker.
  | Delete the file. | Text widget shows the text entered in text input.
  | Click on refresh. | The local storage is intact,
Should be able to generate file when file is deselected. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select generate a file in action & enter file name & select type as csv & enter data. | Generate a file is selected & enter file name & csv is elected as type & data is entered,
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is selected on file picker.
  | Delete the file. | File is generated as csv type.
Should be able to set table page when file is deselected. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Drag & Drop table widget. | Table widget is displayed.
  | Enter more details in the table. | Details are entered.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select set table page & select table name & enter page number. | Set table page is selected & page number is entered.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is selected on file picker.
  | Delete the file. | Table page is set to the page number selected.
Should be able to set variable when file is deselected. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select set variable & enter name. | Set variable is selected in action & name is entered.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is selected on file picker.
  | Delete the file. | Variable is set to the name.
  | Click on inspector. | Variable is displayed.
  | Click on variable. | The name is displayed.
Should be able to unset variable when file is deselected. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select unset variable. | Unset variable is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is selected on file picker.
  | Delete the file. | Variable is unset.
  | Click on inspector. | Variable with 0 items is displayed.
Should be able to control component when file is desecleted. | Create an application. | Application gets created.
  | Drag & drop file picker. | File picker is displayed on the canvas.
  | Select file picker. | File picker is selected.
  | Go to event handler. | Event handler is displayed.
  | Click on +Add Handler. | Event handler with action type as alert is displayed.
  | Select control component & select compnent name & action. | Control component is selected & component name & action is selected.
  | Click on file picker. | Browse is opened.
  | Select a file & click open. | File is being loaded.
  | File is loaded. | File is selected on file picker.
  | Delete the file. | Component is controlled as per action.


</body>
</html>

**Review by sherfin94 — APPROVED** [review]

**Pull request merged**